jupyterhub/utils.py

```diff
@@ -296,6 +299,133 @@ def metrics_authentication(self):
+    if 'user' in kwargs and 'server' in kwargs:
+        user_name = kwargs.pop('user')
+        kwargs['server'] = "{}/{}".format(user_name, kwargs['server'])
+    if len(kwargs) > 1:
+        raise AttributeError("Please specify exactly one filter")
+    if req_scope not in scopes:
+        return False
+    if scopes[req_scope] == Scope.ALL:
+        return True
+    # Apply filters
+    if not kwargs:
+        return False
+    filter_, filter_value = list(kwargs.items())[0]
+    sub_scope = scopes[req_scope]
+    if filter_ not in sub_scope:
+        valid_scope = False
+    else:
+        valid_scope = filter_value in sub_scope[filter_]
+    if not valid_scope and needs_scope_expansion(filter_, filter_value, sub_scope):
+        group_names = sub_scope['group']
+        valid_scope |= check_user_in_expanded_scope(
+            api_handler, filter_value, group_names
+        )
+    return valid_scope
+
+
+def parse_scopes(scope_list):
+    """
+    Parses scopes and filters in something akin to JSON style
+
+    For instance, scope list ["users", "groups!group=foo", "users:servers!server=bar", "users:servers!server=baz"]
+    would lead to scope model
+    {
+       "users":scope.ALL,
+       "users:admin":{
+          "user":[
+             "alice"
+          ]
+       },
+       "users:servers":{
+          "server":[
+             "bar",
+             "baz"
+          ]
+       }
+    }
+    """
+    parsed_scopes = {}
+    for scope in scope_list:
+        base_scope, _, filter_ = scope.partition('!')
+        if not filter_:
+            parsed_scopes[base_scope] = Scope.ALL
+        elif base_scope not in parsed_scopes:
+            parsed_scopes[base_scope] = {}
+        if parsed_scopes[base_scope] != Scope.ALL:
+            key, _, val = filter_.partition('=')
+            if key not in parsed_scopes[base_scope]:
+                parsed_scopes[base_scope][key] = []
+            parsed_scopes[base_scope][key].append(val)
+    return parsed_scopes
+
+
+def needs_scope(scope):
+    """Decorator to restrict access to users or services with the required scope"""
+
+    def scope_decorator(func):
+        @functools.wraps(func)
+        def _auth_func(self, *args, **kwargs):
+            sig = inspect.signature(func)
+            bound_sig = sig.bind(self, *args, **kwargs)
+            bound_sig.apply_defaults()
+            s_kwargs = {}
+            for resource in {'user', 'server', 'group', 'service'}:
+                resource_name = resource + '_name'
+                if resource_name in bound_sig.arguments:
+                    resource_value = bound_sig.arguments[resource_name]
+                    s_kwargs[resource] = resource_value
+            parsed_scopes = parse_scopes(self.scopes)
+            if check_scope(self, scope, parsed_scopes, **s_kwargs):
+                return func(self, *args, **kwargs)
+            else:
+                raise web.HTTPError(403, "Action is not authorized with current scopes")
```

Comment:
Prior to raising, it would be a good idea to add a more detailed `self.log.warning` that includes:

- the path being requested (self.request.path)
- the required scope(s)
- the authorized scope(s)

The error message should also include the required scope (not the current scopes, though, *I think*)